server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,290 @@
+        this.statsMap = new ConcurrentHashMap<>();
+        this.totalStats = new CacheStatsResponse();
+        this.maxDimensionValues = MAX_DIMENSION_VALUES_SETTING.get(settings);
+        this.mode = mode;
+        this.specificCombinations = new HashSet<>();
+    }
+
+    public StatsHolder(List<String> dimensionNames, Settings settings, TrackingMode mode, Set<Set<String>> specificCombinations) {
+        if (!mode.equals(TrackingMode.SPECIFIC_COMBINATIONS)) {
+            logger.warn("Ignoring specific combinations; tracking mode is not set to SPECIFIC_COMBINATIONS");
+        }
+        this.dimensionNames = dimensionNames;
+        this.statsMap = new ConcurrentHashMap<>();
+        this.totalStats = new CacheStatsResponse();
+        this.maxDimensionValues = MAX_DIMENSION_VALUES_SETTING.get(settings);
+        this.mode = mode;
+        for (Set<String> combination : specificCombinations) {
+            assert combination.size() > 0 : "Must have at least one dimension name in the combination to record";
+        }
+        this.specificCombinations = specificCombinations;
+    }
+
+    public StatsHolder(StreamInput in) throws IOException {
+        this.dimensionNames = List.of(in.readStringArray());
+        Map<Key, CacheStatsResponse> readMap = in.readMap(
+            i -> new Key(Set.of(i.readArray(CacheStatsDimension::new, CacheStatsDimension[]::new))),
+            CacheStatsResponse::new
+        );
+        this.statsMap = new ConcurrentHashMap<Key, CacheStatsResponse>(readMap);
+        this.totalStats = new CacheStatsResponse(in);
+        this.maxDimensionValues = in.readVInt();
+        this.mode = in.readEnum(TrackingMode.class);
+        this.specificCombinations = new HashSet<>();
+        int numCombinations = in.readVInt();
+        for (int i = 0; i < numCombinations; i++) {
+            String[] names = in.readStringArray();
+            specificCombinations.add(new HashSet<>(List.of(names)));
+        }
+    }
+
+    public List<String> getDimensionNames() {
+        return dimensionNames;
+    }
+
+    public ConcurrentMap<Key, CacheStatsResponse> getStatsMap() {
+        return statsMap;
+    }
+
+    public CacheStatsResponse getTotalStats() {
+        return totalStats;
+    }
+
+    // For all these increment functions, the dimensions list comes from the key, and contains all dimensions present in dimensionNames.
+    public void incrementHitsByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.hits.inc(amount), 1);
+    }
+
+    public void incrementMissesByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.misses.inc(amount), 1);
+    }
+
+    public void incrementEvictionsByDimensions(List<CacheStatsDimension> dimensions) {
+        internalIncrement(dimensions, (response, amount) -> response.evictions.inc(amount), 1);
+    }
```

Comment:
Could you just pass the `ICacheKey` to the `StatsHolder` and let it extract the values from dimensions?

Reply:
Yes, I like that idea.